src/net.cpp

```diff
@@ -894,40 +893,64 @@ static void EraseLastKElements(
-        const size_t localhost_protect_size{std::max(remaining_tor_slots, localhost_min_protect_size)};
-        EraseLastKElements(vEvictionCandidates, CompareLocalHostTimeConnected, localhost_protect_size,
-                           [](const NodeEvictionCandidate& n) { return n.m_is_local; });
+    // To favorise the diversity of our peer connections, reserve up to half of
+    // these protected spots for onion, localhost and I2P peers, even if they're
+    // not longest uptime overall. This helps protect these higher-latency peers
+    // that tend to be otherwise disadvantaged under our eviction criteria.
+    const size_t initial_size = candidates.size();
+    const size_t total_protect_size = initial_size / 2;
+    struct Net {
+        Network id;
+        size_t count;
+        bool operator<(const Net& rhs) const { return count < rhs.count; }
+    };
+    static constexpr Network localhost{NET_MAX}; // alias localhost to an unused Network enumerator
+    std::vector<Net> networks{{localhost, 0}, {NET_I2P, 0}, {NET_ONION, 0}};
+
+    // Tally the number of eviction candidates per disadvantaged network.
+    for (Net& n : networks) {
+        n.count = std::count_if(candidates.cbegin(), candidates.cend(), [&n](const NodeEvictionCandidate& c) {
+            return n.id == localhost ? c.m_is_local : c.m_network == n.id;
+        });
+    }
+
+    // Find the number of disadvantaged networks from which we have eviction candidates.
+    const auto networks_size = std::count_if(networks.begin(), networks.end(), [](const Net& n) { return n.count; });
+
+    if (networks_size) {
+        // Obtain the minimum number of peers per network to potentially protect.
+        // The base ratio is 1/2 (for the case of peers from only 1 network).
+        const size_t protect_size_per_network{total_protect_size / (2 * networks_size)};
+
+        if (protect_size_per_network) {
+            size_t available_to_protect{protect_size_per_network};
+
+            // Stable-sort networks by ascending candidate count, to allow networks that
+            // have more peers to recover slots not used by networks with fewer peers.
+            std::stable_sort(networks.begin(), networks.end());
+
+            for (const Net& n : networks) {
+                if (n.count == 0) continue;
+
+                // Protect peers of this network, sorted by longest uptime.
+                const size_t peers_to_protect{available_to_protect - (initial_size - candidates.size())};
+                EraseLastKElements(candidates, CompareNodeNetworkTime(n.id, localhost), peers_to_protect,
+                                   [&n](const NodeEvictionCandidate& c) {
+                                       return n.id == localhost ? c.m_is_local : c.m_network == n.id;
+                                   });
+                available_to_protect += protect_size_per_network;
+            }
+        }
```

Comment:
I think the outcome of Case 2 is ideal.  Case 1 protects many I2P peers because that network gets the unused slots from the localhost allocation.  What change would you suggest?  Keep in mind that this is a subcomponent seen in isolation; another 1/4 of the remaining peers will be protected based on their uptime.

Reply:
Hm, maybe we can tally the number of networks that have more than `protect_size_per_network` peers and try to share the leftover slots between them so that Case 1 protects 1 / 11 / 12 instead of 1 / 15 / 8.

Reply:
>  What change would you suggest?

```cpp
void ProtectEvictionCandidatesByRatio(std::vector<NodeEvictionCandidate>& candidates)
{
    // Treat localhost connections as a separate network.
    static constexpr Network NET_LOCALHOST{NET_MAX};

    const std::array<Network, 3> networks{NET_LOCALHOST, NET_I2P, NET_ONION};

    // Protect the half of the remaining nodes which have been connected the longest.
    // This replicates the non-eviction implicit behavior, and precludes attacks that start later.
    // To favorise the diversity of our peer connections, reserve up to half of
    // these protected spots for onion, localhost and I2P peers, even if they're
    // not longest uptime overall. This helps protect these higher-latency peers
    // that tend to be otherwise disadvantaged under our eviction criteria.
    const size_t total_to_protect{candidates.size() / 2};

    // Protect up to 25% of `candidates` due to network. We may protect less if
    // there are no candidates that belong to `networks`.

    const size_t max_protect_due_to_network{total_to_protect / 2};
    size_t num_protected{0};

    while (num_protected < max_protect_due_to_network) {
        const size_t remaining_to_protect{max_protect_due_to_network - num_protected};
        const size_t protect_per_network{std::max(remaining_to_protect / networks.size(), 1UL)};
        bool protected_at_least_one{false};

        for (const Network net : networks) {
            const size_t before = candidates.size();
            EraseLastKElements(candidates,
                               CompareNodeNetworkTime(net, net == NET_LOCALHOST),
                               protect_per_network,
                               [&net](const NodeEvictionCandidate& c) {
                                   return net == NET_LOCALHOST ? c.m_is_local : c.m_network == net;
                               });
            const size_t after = candidates.size();
            if (after < before) {
                protected_at_least_one = true;
                num_protected += before - after;
                if (num_protected >= max_protect_due_to_network) {
                    break;
                }
            }
        }

        if (!protected_at_least_one) {
            break;
        }
    }

    // Protect another 25% to 50% of `candidates` due to uptime/longevity. For example, if
    // 25% were protected due to network then we will protect 25% due to uptime, if
    // 20% were protected due to network then we will protect 30% due to uptime.
    const size_t remaining{total_to_protect - num_protected};
    EraseLastKElements(candidates, ReverseCompareNodeTimeConnected, remaining);
}
```

will result in:
```
(number of peers that belong to localhost, i2p, tor => protected peers from localhost, i2p, tor)
1, 20, 20 => 1, 12, 12
0, 20, 20 => 0, 13, 12
```

Feel free to ignore.

Reply:
This seems like a nice approach, multi-pass with simpler data structures. Will adapt the tests to it to check all the test cases.

Reply:
It looks like we should keep the `Net` struct and `<` sort operator so the result can independent of the order of members in the `networks` array, so a hybrid approach, also adding an `is_local` member to the struct to no longer need `net == localhost` throughout the code.

Reply:
Yes, with the above snippet, the result may depend on the order of elements in the `networks` array, but only up to 1 which I think it ok - after all, if we want to split evenly e.g. 10 slots among 3 networks, some network will get 4 slots and the rest will get 3 slots.

The current code also depends on the order in `networks` because it does "stable" sort, right?

Reply:
The results make more sense when they depend on number of peers by network, e.g. if you have 10 I2P peers and 3 onion ones, best to give first chance at unused slots to the I2P peers so we protect, say, 3 I2P and 2 onions rather than the opposite.  Stable sort only makes a difference if the peer counts are equal. I'll propose this separately in the last commit for feedback.

Reply:
So, if we have 10 I2P peers and 3 onion ones and we want to protect 6 in total, we will protect 3 I2P and 3 onion, this is desired, right?

If we have again 10 I2P and 3 onion and want to protect 5, do we want to protect 3 I2P and 2 onion or 2 I2P and 3 onion? I am not sure. One may argue that protecting one more (but we are talking about just one here) from the rare network (onion) is better because it is easier to find peers from the more popular network (I2P), so the ones from the rare network are more scarce and precious. Or a counter argument could be to keep the proportion - if we have more I2P, then keep more I2P.

Personally, I think it is not so important because it is just 1 peer we are talking about here and I would favor code simplicity over enforcing either approach with more complex code.

Reply:
Yes, this is about tradeoffs. The simplest code was when only a `Network` was passed around all the code without the `is_local` bool everywhere, but that simplicity had a logic encapsulation drawback (perhaps some others), so I rewrote everything and brought back `is_local`. 

I've redone this pull several times because I'd like it to do the right thing, whether networks with more peers get first priority at unused slots or networks with fewer.  If people prefer to have the array order decide independently of the number of peers, it seems we should clearly document that dependency.

I've separated it out to a separate commit, along with the test changes. Am going through everything before pushing the update.

Reply:
My thought is that having separate protection buckets for each of the disadvantaged networks is the essential step here to improving the diversity of our inbound peer connections...and that unused protection slots should be distributed among disadvantaged networks having remaining eviction candidates by descending order of the number of peers.